### ISSUE TITLE
Additional feature, styling and contributions

### ISSUE THREAD

**Issue by asheshambasta** [kmonad/kmonad#266]
Hello! 

I've finally gotten around to making my first contribution to kmonad, and I'd like to make more. In this issue I'd like to discuss this. 

## Additional feature 

When looking through the configuration docs, I found the input device file configuration too specific, at least for systems with just one keyboard (which I'm assuming most systems out there are). I wanted to use the same configuration on my laptop and desktop, but I didn't want to spell out the `*-kbd` file in each case, and hence I've added support for file prefixes and suffixes, in the std. locations: 
- `/dev/input/by-id` 
- `dev/input/by-path`

Here's the diff: https://github.com/asheshambasta/kmonad/commit/01d03d5a9613be428b2c40a56b32a7ea870ffa72#diff-6c00b24c6b5466e1b52b0a2806383cbad8e37fc618d91e0d5a1494762b5902d8R260-R275

@david-janssen if you're ok with this, I'd like to expand upon this (nicer errors, some documentation) and open a PR. 

## Styling 

I've been going through the codebase and I've noticed a few things that I'd like to improve upon:
- Formatting with a std. formatter (stylish-haskell seems the closest to the current style) 
- Fixing HLint warnings 

@david-janssen this will mean some aesthetic changes to the entire set of modules, so I've been refraining from doing that. I'm not sure what your opinion on this is. 

## Future contributions

This is a brilliantly useful project and I'd like to contribute further on this. 

I had already planned to do so (#80) but life got in the way. However, I can put some time into kmonad now and then. 

Also I read through https://github.com/kmonad/kmonad/issues/142, and @david-janssen I hope you're doing better now.

**Comment by david-janssen** [comment]
Hey there,

What an amazingly well organized issue, colour me impressed.

So in regards to your proposed contribution: I think this is a great idea. I like it especially because:
- It adds functionality without breaking any backward-compat
- It makes use of the dispatching mechanisms I had put in place, but hadn't expanded on yet. (i.e. there always was the option of different sources and sinks for the same OS, but up until now it has always been only 1 option per OS, you've added a second option for Linux. So my presience in regards to allowing easy extensibility on this front has paid off, more than a year later :)).

I would very much like this to be a pull-request, however with 1 major change. If you could base it on https://github.com/kmonad/kmonad/tree/keycode-refactor instead? Basically, a lot of the kmonad model, parsing, and invocation-parsing was all mixed up together. I've gone through and separated them out quite a lot. Now the parsing code doesn't do any keyboard initialization anymore, but instead just creates tokens which we pass to model (which then does the initialization). This was a necessary step which will make the following cleanup much easier.

Most of the logic is the same, you'll find it in 'KMonad.App.Parser', mostly. If you have any questions about the changes I'd be happy to answer them.

> I've been going through the codebase and I've noticed a few things that I'd like to improve upon:

Yeah, there's loads of stuff. The pile is high and my energy is limited. I've been focusing on doing the core stuff that has been bottlenecking the other developers (a redesign of the keycode, to make cross-platform support easier, and now separating out model and parser and app into separate modules). There are loads of warts, inconsistencies, name-overlaps, and lack of documentation issues. Any help is super appreciated.

In regards to layouting, I don't even care too much about what style we choose, as long as we make it consistent.

D'oh.. I have to run in 3 minutes, so I'm gonna rush to the end of what I want to say. I'll be back in this issue to finish my thoughts later.

Basically:
- Any help super appreciated
- Let's decide on a style and then try to make it consistent
- Welcome to the club of contributors :)

> Also I read through #142, and @david-janssen I hope you're doing better now.

In regards to that: thank you very much :). It is very much an up and down thing, and currently I'm in a period of things going alright. I try to enjoy the enjoyable, and not lament the unenjoyable too much. It is a practice. The more people who are around and have a good grasp of the code, the more likely KMonad is to persist if I do end up getting bad again.

**Comment by asheshambasta** [comment]
@david-janssen Thanks for the very prompt reply and the welcome! 
Here's how I see your reply: 
- You're ok with this change, as long as its rebased on top of the branch you've mentioned. 
  Should the PR target `master` or this branch?
- You're ok with enforcing a style/formatter: I'd strongly suggest stylish-haskell, which seems the least invasive in this case. 
  I personally prefer brittany, but that will be pretty drastic. 
- You're ok with me fixing some hlint warnings, and other smaller issues. 
  I'm up for doing this, it should be fairly quick and easy. 

I therefore propose:
- I make a PR with this feature alone, rebased on the branch you mentioned 
- I follow up that PR with hlint and styling fixes, which can be merged into the branch you mentioned, and then further work on that branch can continue to happen

Is this OK with you?

**Comment by david-janssen** [comment]
My greek friend is late for our appointment... why am I still surprised by this? :D. Well, now I have time to reply to your issue.

> Should the PR target master or this branch?

No, target `keycode-refactor`. I need to go and to the Windows support, and I'm still waiting on our Mac developer to do the Mac support. Then we'll merge the whole thing into master, but it is so different from master (mainly reorganization) that it's probably easier to integrate your change into the refactor, and then the whole thing into master, than the other way around.

> You're ok with enforcing a style/formatter: 

I am OK, even in favour, but I am not comfortable making this decision quickly. I just scrolled through stylish-haskell, and I'm not too fond of what they do with records. I'd like to give `brittany` a spin and see how that feels. Perhaps we can open a separate issue on this, and discuss a bit further.

> You're ok with me fixing some hlint warnings, and other smaller issues.

Absolutely, that would be great.

> rebased on the branch you mentioned

It might not be an easy rebase, a lot has moved around, so you might have to fix a few conflicts. But yeah, if you can make a pull-request based on `keycode-refactor` for the new key-sources, we'll try and get them integrated right quick. I think they're a great idea. Then the 2nd PR with styling issues also on `keycode-refactor`. I'll try and get all the windows stuff sorted this week, and hopefully we can merge the whole mess into `master` as soon as possible and not exist in this 2-branch limbo for too long.

After that, it is my intention to expand a bit on the unit-tests I've started defining, clean up some of the mess I made during refactoring, make sure the haddocks get built, and then do a new release (of a minor version).

**Comment by asheshambasta** [comment]
@david-janssen I'll write a more detailed reply to your comment soon (tomorrow-ish); and for now I've opened https://github.com/kmonad/kmonad/pull/267 as you suggested.

**Comment by asheshambasta** [comment]
> > You're ok with enforcing a style/formatter:
> 
> I am OK, even in favour, but I am not comfortable making this decision quickly. I just scrolled through stylish-haskell, and I'm not too fond of what they do with records. I'd like to give `brittany` a spin and see how that feels. Perhaps we can open a separate issue on this, and discuss a bit further.

I can strongly recommend brittany: I think while it is highly opinionated, it can really make the code look beautiful. 
We use Brittany at work: and while it has some caveats: 
- No support for arrows
- Slight wonkiness with TypeOperators and such (more recent versions) 
These can be overcome by adding a simple 
```
-- brittany-disable-next-binding
```

However, keep in mind that the changes Brittany will make will be pretty large, and I'd suggest going about that in a different PR, meant just for styling. 

> 
> > You're ok with me fixing some hlint warnings, and other smaller issues.
> 
> Absolutely, that would be great.
> 

I'll get to this once #267 is merged. 

> > rebased on the branch you mentioned
> 
> It might not be an easy rebase, a lot has moved around, so you might have to fix a few conflicts. But yeah, if you can make a ... 

See #267. 

> After that, it is my intention to expand a bit on the unit-tests I've started defining, clean up some of the mess I made during refactoring, make sure the haddocks get built, and then do a new release (of a minor version).

Another issue I notice, that at least on my branch, the nix-build seems to be broken (missing hspec, which should be an easy fix), and it is likely this comes from the rebase: so it might be also present in your current branch. 

---

> In regards to that: thank you very much :). It is very much an up and down thing, and currently I'm in a period of things going alright. I try to enjoy the enjoyable, and not lament the unenjoyable too much. It is a practice. The more people who are around and have a good grasp of the code, the more likely KMonad is to persist if I do end up getting bad again.

Well, I hope you stay well for a long time and KMonad doesn't need to persist without you. Health is a delicate thing, and given we're all faced with sitting jobs, it is important we take care of it.

**Comment by david-janssen** [comment]
In regards to styling, I am starting to have some doubts about using a code-prettifier tool. I remember that when I used stylish-haskell, sometimes I'd end up struggling against the tool, and I'd end up writing code I was less happy with just to make it conform better to some automatic tool. Additionally, we are adding an extra point of failure. Seeing as the core `brittany` developer has also gone AWOL for a bit.

The other option would be to simply document our preferences for styling in some documentation, and then kindly ask people to follow it as much as possible, but to violate it when the violation leads to greater clarity. I think for such a small hobby-project written by a handful of guys, that is probably the friendlier and more pleasant solution.

If you can convince both @slotThe and @thoelze1 that `brittany` is a great idea, I'm happy to give it a shot.

> Another issue I notice, that at least on my branch, the nix-build seems to be broken (missing hspec, which should be an easy fix), and it is likely this comes from the rebase: so it might be also present in your current branch.

Ah, well spotted. I was going to dig into making sure all the building was correct on all the platforms once we got around to preparing the next version bump (after the refactor is done). After that, it'd probably be a good idea to set up some automated build-tests to catch these little issues automatically.

> Well, I hope you stay well for a long time and KMonad doesn't need to persist without you. 

Anicca, my friend :-). Nothing fabricated will escape the grinding wheels of time.

**Comment by slotThe** [comment]
> The other option would be to simply document our preferences for
> styling in some documentation, and then kindly ask people to follow it
> as much as possible, but to violate it when the violation leads to
> greater clarity. I think for such a small hobby-project written by a
> handful of guys, that is probably the friendlier and more pleasant
> solution.

I think this would be a good solution, though it would perhaps be a bit
harder on first-time contributors who only want to fix/add a small
thing.  What a few projects do in that case is to merge the pull request
anyways and then push fixup commits themselves.  If the person turns out
to be a repeated contributor then we can always tell them to look at the
style guide then and there.

On Tue, May 25 2021 09:34, David Janssen wrote:
> If you can convince both @slotThe and @thoelze1 that `brittany` is a
> great idea, I'm happy to give it a shot.

I have to say that I'm not a big fan of auto-formatters in general.
They mostly destroy alignment and... I like pretty things :) Makes many
things much easier to read in my opinion.

**Comment by david-janssen** [comment]
> > If you can convince both @slotThe and @thoelze1 that brittany is a great idea, I'm happy to give it a shot.

Looking back on this, I actually think it was a bit unfair of me to hide behind them, and I apologize. I am still getting my bearings 'making decisions' for a project, and I struggle with wanting to please everybody. This is my own personal struggle, because as the user- and contributor-base grows, the more impossible this becomes.

I will still let the ideas percolate a bit and make a decision about this when I am rushing a bit less. I completely get your point about the benefit of automation, but I also have experience with 'struggling against' autoformatters.

I think I will give both `brittany` and `stylish-haskell` a serious try when I find some time after all the actual work required to get things working. If I can get either one to make things prettier without getting in the way, I will choose one of them. Otherwise, I will start a style-guide.

**Comment by david-janssen** [comment]
> We have some experience setting up GH actions to build with nix.

Ooh, when the time comes your counsel would be much appreciated.
